mautic/mautic
# Code Mode

# Caution: Work in progress

| Q | A |
| --- | --- |
| Bug fix? | N |
| New feature? | Y |
| Related user documentation PR URL | TODO |
| Related developer documentation PR URL | / |
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/2370, https://github.com/mautic/mautic/issues/2368, https://github.com/mautic/mautic/issues/2737 |
| BC breaks? | N |
| Deprecations? | N |
#### Description:

There has been many emotional discussions about Froala editor modifying the code. This PR creates the "Code Mode" for each email/page and removes the Content Tab. All content modifications can be done only via the Builder. If a theme is selected, the builder looks the same as before. However, if the Code Mode is selected, the builder looks like this:
![mautic-code-mode-builder](https://cloud.githubusercontent.com/assets/1235442/19193653/296a16f4-8cab-11e6-9591-ea2a427935b3.png)
The preview will refresh automatically every 10 seconds if there was a change in the code.
##### The workflows can go like this:
1. If user wants to create an email/page and use his own HTML, he goes directly to the Code Mode, paste the HTML in the Builder's code editor and save.
2. If user wants to use a theme, he select a theme and can edit the content in the Builder.
3. If user wants to edit HTML of an email/page created by a theme, he can switch to the Code Mode, but NOT BACK.
4. The same works for BC. If user wants to convert email/page created before the Code Mode was available, he can switch to Code Mode now with old emails/pages, but NOT BACK.
##### Known issues:
1. The features built into Froala won't be easily available in the Code editor (video, shortcode btn, shortcode dropdown when `{` is pressed, Dynamic email content button, images...)
2. Video and dynamic email content is not built into the Builder editors either.
#### Steps to test this PR:
1. Try to create an email with Code Mode with custom HTML. Check the email preview.
2. Try to create an email from a predefined theme, edit in Builder, save. Check the email preview.
3. Try to switch the email created from 2. to the Code Mode, edit the code, save. Check the email preview.
4. Edit all steps above with pages as well.
